src/NewSkin.h

```diff
@@ -0,0 +1,433 @@
+	Actor* get_player_tap(size_t type, size_t pn, double beat);
+	Actor* get_player_optional_tap(size_t type, size_t pn, double beat);
+	void get_hold_render_data(TapNoteSubType sub_type,
+		NotePlayerizeMode playerize_mode, size_t pn, bool active, bool reverse,
+		double quantization, double beat, QuantizedHoldRenderData& data);
+	double get_width() { return m_width; }
+	double get_padding() { return m_padding; }
+	bool supports_masking()
+	{
+		return !(m_hold_player_masks.empty() || m_hold_reverse_player_masks.empty());
+	}
+	bool load_holds_from_lua(lua_State* L, int index,
+		std::vector<std::vector<QuantizedHold> >& holder,
+		std::string const& holds_name,
+		NewSkinLoader const* load_skin, std::string& insanity_diagnosis);
+	bool load_texs_from_lua(lua_State* L, int index,
+		std::vector<RageTexture*>& dest,
+		std::string const& texs_name,
+		NewSkinLoader const* load_skin, std::string& insanity_diagnosis);
+	bool load_from_lua(lua_State* L, int index, NewSkinLoader const* load_skin,
+		std::string& insanity_diagnosis);
+	void vivid_operation(bool vivid)
+	{
+		for(auto&& tap : m_taps)
+		{
+			tap.m_vivid= vivid;
+		}
+		for(auto&& tap : m_optional_taps)
+		{
+			if(tap != nullptr)
+			{
+				tap->m_vivid= vivid;
+			}
+		}
+		for(auto&& subtype : m_holds)
+		{
+			for(auto&& action : subtype)
+			{
+				action.m_vivid= vivid;
+			}
+		}
+		for(auto&& subtype : m_reverse_holds)
+		{
+			for(auto&& action : subtype)
+			{
+				action.m_vivid= vivid;
+			}
+		}
+	}
+	void clear_optionals()
+	{
+		for(auto&& tap : m_optional_taps)
+		{
+			if(tap != nullptr)
+			{
+				SAFE_DELETE(tap);
+			}
+		}
+	}
+	NewSkinColumn()
+		:m_optional_taps(NUM_NewSkinTapOptionalPart, nullptr)
+	{}
+	~NewSkinColumn()
+	{
+		clear_optionals();
+	}
+private:
+	// m_taps is indexed by NewSkinTapPart.
+	std::vector<QuantizedTap> m_taps;
+	// m_optional_taps is indexed by NewSkinTapOptionalPart.
+	// If an entry is null, the skin doesn't use that part.
+	std::vector<QuantizedTap*> m_optional_taps;
+	// Dimensions of m_holds:
+	// note subtype, active/inactive.
+	std::vector<std::vector<QuantizedHold> > m_holds;
+	std::vector<std::vector<QuantizedHold> > m_reverse_holds;
+	// m_hold_player_masks is indexed by note subtype.
+	std::vector<RageTexture*> m_hold_player_masks;
+	std::vector<RageTexture*> m_hold_reverse_player_masks;
+	// m_rotation_factors stores the amount to rotate each NSTP.
+	// So the noteskin can set taps to rotate 90 degrees in this column and
+	// mines to rotate 0, and taps will be rotated and mines won't.
+	std::vector<double> m_rotations;
+	double m_width;
+	double m_padding;
+};
+
+struct NewSkinLayer
+{
+	bool load_from_lua(lua_State* L, int index, size_t columns, std::string& insanity_diagnosis);
+	// The actors are public so that the NewFieldColumns can go through and
+	// take ownership of the actors after loading.
+	std::vector<Actor*> m_actors;
+};
+
+struct NewSkinData
+{
+	static const size_t max_columns= 256;
```

Comment:
I needed a number to detect someone screwing up and putting way to many columns in on accident, but allow future expansion.  256 seemed like a fair choice.  A smaller number doesn't really give more benefit.